Universal Recorder Overhaul

Davidj361 opened the pull request:
:cl: Davidj361/MrSavage
tweak: Overhauled how the universal recorder records and plays, it should now seem like the original person they recorded is there actually saying it. (Get ready for some trickery!)
add: The universal recorder now has a menu.
add: Can view the transcript without having to print.
add: Can play/pause the universal recorder. (You cannot over-record the tape still)
add: The universal recorder can toggle announcements for while being played that it starts/ends playing/recording
add: Player able to alt-click the universal recorder so they can toggle the recording of a universal recorder while it's in their pocket
add: Can toggle the universal recorder recording the communications from radios
add: Seekbar for the universal recorder
add: Re-added icons of objects/items to browser menus. (No idea why this was commented out in the first place)
wip: Map makers can make pre-recorded tapes
/:cl:
~~add: The universal recorder can loop~~

The universal recorder needed some more potential as it was seen as a fluff item. I believe the universal recorder should be equivalent to Kevin using a voice recorder in Home Alone for shenanigans.

Davidj361 committed:
Alt-click universal recorder & toggle comms

Made the player able to alt-click the universal recorder so they can toggle the recording of a universal recorder while it's in their pocket.
Added a verb to toggle the universal recorder recording the communications from radios.

Davidj361 committed:
Added loop, announce, and better play features to recorder.

- The universal recorder can loop
- The universal recorder can toggle announcements for while being played that it starts/ends playing/recording
- Overhauled how the universal recorder records and plays, it should now seem like the original person they recorded is there actually saying it. (get ready for some trickery!)

Davidj361 committed:
Added loop, announce, and better play features to recorder.

- The universal recorder can loop
- The universal recorder can toggle announcements for while being played that it starts/ends playing/recording
- Overhauled how the universal recorder records and plays, it should now seem like the original person they recorded is there actually saying it. (get ready for some trickery!)

ShadowDeath6 commented:
Do radios hear these? Loop sounds like a bad idea if you can spam radio with them infinitely.

Davidj361 commented:
Is the AI unable to locate the radio being used?

Davidj361 committed:
Merge branch 'universalrecorder' of https://github.com/Davidj361/tgstation into universalrecorder

Davidj361 committed:
Merge branch 'universalrecorder' of https://github.com/Davidj361/tgstation into universalrecorder

RemieRichards commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -115,7 +148,7 @@
 		update_icon()
 		mytape.timestamp += mytape.used_capacity
-		mytape.storedinfo += "\[[time2text(mytape.used_capacity * 10,"mm:ss")]\] Recording started."
+		mytape.addinfo("Recording started.", , , , , , TRUE)
```
used named args instead of many ,,,,,

RemieRichards commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -165,28 +198,35 @@
 	update_icon()
-	usr << "<span class='notice'>Playing started.</span>"
+
+	if (announce) say("Playing started.")
```
an if's effects should not be on the same line.

Davidj361 replied:
Why not? It works.

RemieRichards replied:
We have standards.

Davidj361 committed:
God damn I'm tired

RemieRichards commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -165,28 +198,35 @@
 	var/max = mytape.max_capacity
 	for(var/i = 1, used < max, sleep(10 * playsleepseconds))
-		if(!mytape)
+		if(!mytape || playing == 0 || mytape.storedinfo.len < i)
```
is playing being used as a bool here? if so ```!playing```

RemieRichards commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -165,28 +198,35 @@
-		say(mytape.storedinfo[i])
+		var/x = mytape.storedinfo[i]
+		// Burrowed code from send_speech
+<<<<<<< HEAD
```
Great. fix your merge :)

RemieRichards commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -165,28 +198,35 @@
 			say("Skipping [playsleepseconds] seconds of silence")
 			playsleepseconds = 1
 		i++
+	if (announce) say("End of recording.")
```
same as above.

Dorsisdwarf commented:
There should be an option to have it playback as the device, so you can avoid confusion if you want. Otherwise recordings involving yourself could be a pain

Davidj361 commented:
That's why I thought announcements of what the tape is doing should explain that. Otherwise, I suppose I could add timestamps at the end of the messages to make it even more clear.

Davidj361 commented:
I just noticed there's a bug over the radios. Anyhow treat this as broken for now until I fix it.

WJohn commented:
Will it be possible to add pre-recorded messages from the map maker? There's only so much I can get away with when writing things down on sheets of paper and still make it seem realistic (sudden cutoffs and the like are not usable, no one suddenly puts a dash at the end or keeps writing when someone has barged in the room at a point you can reasonably figure out what happened).

Davidj361 commented:
Can you write what it would look like and describing the situation?

WJohn commented:
Check the bbcode segments here https://tgstation13.org/wiki/Guide_to_paperwork#

I'm not sure if it's possible to add your own functions to bbcode, but if possible I'd love to have one like [wait5] or whatever that gives a 5 decisecond pause between the previous line and the next. Might also need a [print] or something to illustrate the end of a sentence that you want spoken at that moment.

For example:

We've recovered the sample from the asteroid successfully.[print][wait30]It's giving off highly anomalous readings...[print][wait20]Wait... something's happening.[print]

I dunno, just throwing ideas around, you can handle it however you want, but I'd love to see this added. Holotapes would be so cool to find.

Davidj361 committed:
Removed Loop, fixed the tape recorder to work over radio, added pause
- Removed the loop until AI has the ability to see where the tape recorder is playing over the radio
- Changed addinfo to addInfo, added addInfoAnnounce for tape recorder announcements
- The tape player now properly masks itself as the individual when played over the radio
- The player can now pause the tape playing (you can't over-record the tape)

Davidj361 committed:
Removed Loop, fixed the tape recorder to work over radio, added pause
- Removed the loop until AI has the ability to see where the tape recorder is playing over the radio
- Changed addinfo to addInfo, added addInfoAnnounce for tape recorder announcements
- The tape player now properly masks itself as the individual when played over the radio
- The player can now pause the tape playing (you can't over-record the tape)

JJRcop commented:
uh oh

MrPerson commented:
I don't want to be an asshole but this is going to be a nightmare to maintain. You know what everybody loves? Lists with magic number indices in a specific, illogical, undocumented order, some of which are null for seemingly no reason. Very annoying to tweak or expand upon this as a result.

Davidj361 commented:
I don't understand what you mean.

Davidj361 commented:
Noted, I will try to make it more clear.

optimumtact commented:
Use defines with readable names for the index

#define thisindexkey 8

x[thisindexkey]

Davidj361 commented:
yeah, I'm going to finish this up later this week. Still have classes & homework that I have to do instead. Also I want to make a TGUI seek menu, but the damn thing isn't compilable on windows so I'll have to compile it in my virtualbox linux (vm).

Davidj361 committed:
Fixed indexing, viewability option
- Made the indexing for stored info use keys instead of random number indexes
- If announce mode is on, the tape recorder talks in purple and italics

Davidj361 committed:
Merge branch 'master' into universalrecorder

Davidj361 committed:
Seekmenu, refactored code
- Added a menu for the tape recorder, activate by using it in your hand. It lets you do everything except eject. You can also seek and it has a seek bar.
- Tape recorder makes a noise when winding while it's playing
- made timestamps into timestampsInSeconds to make it more obvious
- Changed some other code around

Davidj361 commented:
Forgot to add in the commit that I made it more obvious in announce mode that the tape recorder is speaking.
The last feature that needs to be implemented is the thing that makes it easier for map makers to make tapes with conversations.

Davidj361 committed:
Uncommented
I have no idea why these were commented out in the first place, there weren't any comments on git or the code explaining why there were commented out. Basically functionality was taken out of the menus having a nice little icon in the title bar. So I put it back in.

Davidj361 committed:
Icons, uses browser datum, view transcript,variables, bug fixes, auto-refresh, more
- Had to export the gifs from a dmi file because they wouldn't be animated in the browser menu
- Browser
    - Uses browser datum now so the menu looks prettier
    - Has the icon enlarged in the menu and becomes animated when recording/playing
    - Has a visual seek bar with text over it
    - View Transcript option added so you don't have to print
    - auto-refreshes depending on the actions happening with the tape recorder
- Changed the recording stopped & played to *an audible cut...* so less spam
- added announce function so colors remain
- there's a span for abductor so consistence in color remains
- function for creating paper data

Davidj361 committed:
Merge branch 'master' into universalrecorder

Davidj361 committed:
Added map-maker functionality
Map-makers now have a simple markup language to make prerecorded messages on tapes. The functionality is in place, but the interface for the map-makers is not there yet as I have to re-confirm with how the interface should be.

Davidj361 committed:
Merge branch 'master' into universalrecorder

Davidj361 commented:
Still got more work to do on this, all these might or not be implemented:

- add seek buttons for play position
- Make seek bar actually track every second
- Time isn't updated while recording when nothing is spoke
- Steals focus
- make setDialogue multiline
- first use/reconnecting makes the tape recorder steal focus when menu is open
- toggle skip silence
- possibly use audible_message or visible_message

XDTM commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -287,4 +602,55 @@
+		else
+			if (length(re.group[1]) != 0) // Is it a blank []?
+				if (re.group[1] == ";") // Quick announcement?
+					speaker = TAPE_RECORDER_NAME
```
Should probably be `speaker = name` instead of a constant, in case the tape recorder changes name for some reason.

Davidj361 replied:
I wanted to access the name of /obj/item/device/taperecorder from /obj/item/device/tape/verb/setDialogue(txt as text).
/obj/item/device/taperecorder has a variable of an instance of /obj/item/device/tap, so I thought the easiest way would be to use a #define without having to initialize a /obj/item/device/taperecorder for its name variable.

XDTM commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -287,4 +602,55 @@
+					speaker = re2.group[1]
+			else
+				wait = 1
+				speaker = TAPE_RECORDER_NAME
```
same here

KorPhaeron commented:
Stale, bug me when it has progress
